text/0000-extensible-enums-and-structs.md

```diff
@@ -0,0 +1,125 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Introduce a notation for declaring an enum to be *extensible*, meaning
+that more variants or fields may be added in the future. This notation
+does not affect the use of the enum within the current crate, but it
+prevents downstream crates from employing features (such as exhaustive
+matches or struct literals) that would fail to compile if new
+variants/fields were added.
+
+# Motivation
+
+Imagine that one is writing a public-facing API that includes an
+`Error` enum (this exact scenario arises in libstd with the I/O error
+enum):
+
+```rust
+pub enum Error {
+    NoSuchFile,
+    NoPermissions,
+}
+```
+
+Providing an enum listing out the possible sources of error is a very
+convenient API for your users, because they can be use a `match`
+statement to easily identify and handle particular kinds of
+errors. Unfortunately, it's rather limiting on the future evolution of
+your library: if you wish to add a third kind of error in the future,
+you will be potentially breaking downstream code. After all, it is
+possible that some user was doing an exhaustive match against all
+possible sources of error:
+
+```rust
+match error {
+    Error::NoSuchFile => ...,
+    Error::NoPermissions => ...,
+}
+```
+
+To resolve this dilemna, this RFC proposes permitting enums to be declared
+as *extensible*:
+
+```rust
+pub enum Error {
+    NoSuchFile,
+    NoPermissions,
+    ..
+}
+```
+
+Extensible enums from other crates can never be exhaustively
+matched. This means that a downstream user attempting to match against
+an error from your library would have to include a wildcard option:
+
+```rust
+// from an external crate
+match error {
+    Error::NoSuchFile => ...,
+    Error::NoPermissions => ...,
+    _ => ...,
+}
+```
+
+Due to the presence of this wildcard, you can safely add variants
+without fear of breaking source compatibility with downstream clients.
+
+Note that extensibility is ignored within the current
+crate. Therefore, your library may internally write a match that
+exhaustively covers all sources of error. You don't need to worry
+about source compatibility with yourself, after all.
+
+# Detailed design
+
+Enum grammar is extended to permit a list of variants to be terminated
+with `..`. An enum declared with `..` is considered *extensible*.
```

Comment:
+1 for `...`
On Jan 29, 2015 7:50 AM, "Nick Cameron" <notifications@github.com> wrote:

> In text/0000-extensible-enums-and-structs.md
> <https://github.com/rust-lang/rfcs/pull/757#discussion_r23735568>:
>
> > +    _ => ...,
> > +}
> > +```
> > +
> > +Due to the presence of this wildcard, you can safely add variants
> > +without fear of breaking source compatibility with downstream clients.
> > +
> > +Note that extensibility is ignored within the current
> > +crate. Therefore, your library may internally write a match that
> > +exhaustively covers all sources of error. You don't need to worry
> > +about source compatibility with yourself, after all.
> > +
> > +# Detailed design
> > +
> > +Enum grammar is extended to permit a list of variants to be terminated
> > +with `..`. An enum declared with `..` is considered *extensible*.
>
> Excuse the bikeshedding, but I'd like to propose ... rather than ..
> because a) .. is much more heavily overloaded in Rust already than ...
> and b) ... actually means 'elided things' in English writing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/757/files#r23735568>.
>